src/orange/univ/SimpleUnitTracker.hh

```diff
@@ -202,26 +206,29 @@ CELER_FUNCTION auto
 SimpleUnitTracker::cross_boundary(LocalState const& state) const -> Initialization
 {
     CELER_EXPECT(state.surface && state.volume);
-    detail::SenseCalculator calc_senses(
-        this->make_surface_visitor(), state.pos, state.temp_sense);
 
     detail::OnLocalSurface on_surface;
-    auto is_inside = [this, &state, &calc_senses, &on_surface](
-                         LocalVolumeId const& id) -> bool {
+    auto is_inside
+        = [this, &state, &on_surface](LocalVolumeId const& id) -> bool {
+        detail::LazySenseCalculator calc_senses(this->make_surface_visitor(),
+                                                state.pos,
+                                                state.temp_sense,
+                                                state.temp_sense_mod);
         if (id == state.volume)
         {
             // Cannot cross surface into the same volume
             return false;
         }
 
         VolumeView vol = this->make_local_volume(id);
-        auto logic_state
-            = calc_senses(vol, detail::find_face(vol, state.surface));
-
-        if (detail::LogicEvaluator(vol.logic())(logic_state.senses))
+        auto on_face = detail::find_face(vol, state.surface);
+        auto bind_calc_sense = [&](FaceId face_id) {
+            return calc_senses(vol, face_id, on_face);
+        };
```

Comment:
Why not have the LazySenseCalculator just take a `face_id` since it's constructed locally where you already have `vol` and `on_face`? Alternatively why not construct the lazy calculator outside `cross_boundary`?

Reply:
We could change the API and pass `vol` and `on_face` when constructing the LSC.  We have to build the LSC inside the `is_inside` lambda because of the handling logic for `on_face`, which must be reset every time. 

This would make the LSC instance usable only for a specific volume, however, given that we need to insert an ´invalidate_cache` if we're reusing it for another volume ( as is done inside `initialize`),  this might be better as this is quite easy to misuse as it is now.

Reply:
Yeah, I think that's a better approach. The original sense calculator had track/universe as "constant" data as a convenience (the constructor does nothing) and transformed one volume to one list of results; there's no reason we can't change the semantics for this use case.